extended `LAMMPS` commands support

cz4rs opened the pull request:
Make it possible to define multiple regions (and the size of simulation box will be adjusted accordingly). This is currently rectricted to `fcc` lattice style only.

Notable changes:
- use `create_atoms` to fill the region with atoms of specific type
- use `mass` to assign mass to atom type
- use `velocity` to assign temperature to atom type

See `in.lb` deck and the screenshots below for practical demonstration.

You can use `paraview --script utils/paraview_process_data.py` to speed up analysis of produced data. It will try to load all `dump*`, `domain_act*` and `domain_lb*` data files from your current directory and do some basic ParaView visualization.

---

TODO (mostly optional):

- [ ] run the simulations for existing decks in CI to ensure no parsing or runtime errors
- [ ] deduce the padding width for filenames from number of steps instead of using hardcoded default
- [ ] improve compatibility with LAMMPS (to be able to run the decks with LAMMPS for comparison)
- [ ] improve `group` command support
- [x] improve communication statistics

---

Multiple regions with different atom types. Note that when the regions overlap, random type will be assigned:
![image](https://github.com/user-attachments/assets/4afaa255-724e-4d97-b714-4fec2aae46f9)

Multiple regions with different velocities:
![image](https://github.com/user-attachments/assets/9199498f-7359-453d-a712-737d2a03199d)

Types after 100 steps:
![image](https://github.com/user-attachments/assets/316abdb5-e984-4d53-b8c0-30f31840ab80)

Velocities after 100 steps:
![image](https://github.com/user-attachments/assets/e9963cf2-e678-4465-a020-7fb4e46887dd)

Types after 1000 steps:
![image](https://github.com/user-attachments/assets/701ce523-327e-48e4-8303-ef25669df706)

cz4rs committed:
Print number of particles sent and received

cz4rs committed:
Add basic support for `group` command

cz4rs committed:
Update .gitignore

cz4rs committed:
Use a struct for storing region coordinates

cz4rs committed:
Add ParaView script for automated data processing

cz4rs committed:
Allow boxes with non-zero offset

cz4rs committed:
Use two atom types

cz4rs committed:
Use different velocities depending on atom type

cz4rs committed:
Support multiple regions

cz4rs committed:
Decide atom type based on region

cz4rs committed:
Add customized input deck

cz4rs committed:
Fix compatibility with original decks

cz4rs committed:
Fix formatting

cz4rs committed:
Use 4 digits in data files by default

cz4rs committed:
Add missing default

cz4rs commented:
@streeve This is remaining part of our work on CabanaMD. Marking this as a draft, as there are some limitations (mostly being restricted to `fcc` lattice style only) and in progress bits.
Would you be interested in getting this merged as well?

streeve commented:
> This is remaining part of our work on CabanaMD. Marking this as a draft, as there are some limitations (mostly being restricted to `fcc` lattice style only) and in progress bits. Would you be interested in getting this merged as well?

Yes definitely, I will take a look

streeve commented on paraview_process_data.py:
This is definitely useful - maybe we make a `scripts` or `utils` folder?

cz4rs replied:
Moved to `utils`.

streeve commented on src/cabanamd_impl.h:
```diff
@@ -328,6 +328,7 @@ void CbnMD<t_System, t_Neighbor>::run()
 
             // Exchange atoms across MPI ranks
             comm_timer.reset();
+            log( std::cout, "Comm::exchange()\tstep ", step );
```
Should this be unified with the rest of the output?

cz4rs replied:
Removed (non-significant).

streeve commented on src/comm_mpi_impl.h:
```diff
@@ -268,6 +271,11 @@ void Comm<t_System>::exchange()
 
     N_local = N_local + N_total_recv - N_total_send;
 
+    std::cout << '[' << proc_rank << "] "
```
Same here - always better to have more output, but I doubt we always want to output this

cz4rs replied:
Yes, definitely. I will do some testing and polishing to make sure that the information is actually useful.

cz4rs replied:
Relevant `cabanaMD.out` fragment after integrating this with the other statistics:
```
#Timestep Temperature PotE ETot Time Atomsteps/s LBImbalance Exchanged
0   13.038466   -4.081539   15.472248   0.00    0.00e+00    -nan    -   
10  11.276282   -1.442460   15.468581   0.52    9.67e+04    -nan    -   
20  10.717567   -0.607034   15.466101   1.13    8.11e+04    7.55e-01    1057
30  10.846505   -0.798659   15.467845   1.87    6.77e+04    7.55e-01    -   
40  10.953565   -0.960041   15.467020   2.73    5.80e+04    7.14e-01    1011
50  11.091998   -1.167610   15.467059   3.15    1.19e+05    7.14e-01    -   
60  11.050713   -1.105665   15.467089   3.65    1.02e+05    6.50e-01    704 
70  11.114292   -1.200594   15.467509   4.00    1.41e+05    6.50e-01    -   
80  11.153192   -1.258347   15.468094   4.40    1.26e+05    5.65e-01    639 
90  11.169153   -1.281515   15.468864   4.66    1.89e+05    5.65e-01    -   
100 11.207249   -1.339408   15.468103   4.96    1.66e+05    4.70e-01    555 
```

`Exchanged` column will show up only when Load Balancing is enabled.

cz4rs replied:
I went with total values since we don't seem to need a breakdown per rank.

streeve commented on src/inputFile_impl.h:
```diff
@@ -270,12 +273,40 @@ void InputFile<t_System>::check_lammps_command( std::string line,
         known = true;
-        int type = std::stoi( words.at( 1 ) ) - 1;
+        int type = std::stoi( words.at( 1 ) ) -
+                   1; // this is annoying because all other commands have to do
```
I have no problem updating this more broadly if it makes things easier

cz4rs replied:
Let's leave it as it is, I've removed the comment.

streeve commented on input/in.darma:
As someone not very familiar with darma, is it more clear to call this `load_balance` or `lb`?

cz4rs replied:
Renamed to `in.lj`.

cz4rs committed:
Allow empty regions

cz4rs committed:
Rename load balancing deck

cz4rs committed:
Move ParaView script to `utils`

cz4rs committed:
Drop redundant comments

cz4rs commented:
@cwschilly This should help with modifying the load balancing test case.

@streeve Thanks for the review. Leaving this as a draft for now, I will fix the output and possibly some other minor issues next week :crossed_fingers:

streeve commented:
@cz4rs just a note that I updated the formatting to 14 here to match Cabana (#117) and in the near future we'll plan to bump that again to match Kokkos at 16

cz4rs commented:
> @cz4rs just a note that I updated the formatting to 14 here to match Cabana (#117) and in the near future we'll plan to bump that again to match Kokkos at 16

Alright, I'll rebase this as needed.

cz4rs committed:
Fix formatting

cz4rs committed:
Extract summary printing into a method

- print imbalance on initial step (will be `-nan`)
- use tabs as separators for easier reading

cz4rs committed:
Make communication statistics consistent with other output

cz4rs commented:
@streeve I have addressed remaining comments. This is also rebased and re-formatted.

streeve commented on src/cabanamd_impl.h:
```diff
@@ -681,3 +646,37 @@ void CbnMD<t_System, t_Neighbor>::check_correctness( int step )
+        {
+            log( out, "\n#Timestep Temperature PotE ETot Time Atomsteps/s "
+#ifdef CabanaMD_ENABLE_LB
+                      "LBImbalance Exchanged"
```
```suggestion
                      "LB-Imbalance MPI-Exchanged"
```

streeve commented on input/in.lb:
```diff
@@ -0,0 +1,41 @@
+# Example using Lennard-Jones potential
```
```suggestion
# Example for exercising load balancing (not intended to be physical)
```

streeve reviewed (commented):
This looks great - couple of wording suggestions, but I think it's ready to merge after that

cz4rs committed:
Apply suggestions from code review

Co-authored-by: Sam Reeve <6740307+streeve@users.noreply.github.com>
